airbyte-integrations/connectors/source-hubspot/source_hubspot/components.py

```diff
@@ -0,0 +1,48 @@
+#
+# Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+#
+
+from typing import Any, Dict, Optional
+
+from airbyte_cdk.sources.declarative.transformations import RecordTransformation
+from airbyte_cdk.sources.types import Config, StreamSlice, StreamState
+
+
+class NewtoLegacyFieldTransformation(RecordTransformation):
+    """
+    Implements a custom transformation which adds the legacy field equivalent of v2 fields for streams which contain Deals and Contacts entities.
+
+    This custom implmentation was developed in lieu of the AddFields component due to the dynamic-nature of the record properties for the HubSpot source. Each
+
+    For example:
+    hs_v2_date_exited_{stage_id} -> hs_date_exited_{stage_id} where {stage_id} is a user-generated value
+    """
+
+    def __init__(self, field_mapping: Dict[str, str]) -> None:
+        self._field_mapping = field_mapping
+
+    def transform(
+        self,
+        record_or_schema: Dict[str, Any],
+        config: Optional[Config] = None,
+        stream_state: Optional[StreamState] = None,
+        stream_slice: Optional[StreamSlice] = None,
+    ) -> None:
+        """
+        Transform a record in place by adding fields directly to the record by manipulating the injected fields into a legacy field to avoid breaking syncs.
+
+        :param record: The input record to be transformed
+        """
+        for field, value in list(record_or_schema.get("properties", record_or_schema).items()):
+            for legacy_field, new_field in self._field_mapping.items():
+                if new_field in field:
+                    transformed_field = field.replace(new_field, legacy_field)
+
+                    if legacy_field == "hs_lifecyclestage_" and not transformed_field.endswith("_date"):
+                        transformed_field += "_date"
+
+                    if record_or_schema.get("properties") is not None:
+                        if record_or_schema["properties"].get(transformed_field) is None:
+                            record_or_schema["properties"][transformed_field] = value
```

Comment:
I think the main thing that feels weird about this component/logic is that we're combining schema modification and record transformation into the same component. Its a little confusing in that regard, but not a dealbreaker.

But the reason why I'm a little opposed about this is that when we do move Hubspot to low-code, this custom component will only ever get invoked at runtime on records since thats how the low-code framework uses transformations. And we're going to have to reimplement this (or maybe we get it for free depending how the dynamic schema component is designed). So given that schema transformation should eventually be removed from the component, it makes me think that we should just implement the schema part separately from the custom component.

Again, I know we have a time crunch to get this done, so ultimately this is just a nit, but this was my main thought that kept popping up in my head during the review.